### PR TITLE
Support preCICE Waveform API

### PR DESCRIPTION
**Pull request by BenjaminRodenberg**
This is not necessarily breaking, because the waveform API was already introduced in preCICE v2.4.0.

### CODE REVIEW

**Comment by BenjaminRodenberg** [comment]
Everything except the spack tests are working here. (I currently do not have the time to look deeper into spack)

I will now also create a breaking PR for the 3.0.0 release of preCICE.

**Comment by IshaanDesai** [comment]
Can we sync this with the latest develop to ensure that when we plan to merge this we would not have excessive conflicts?

**Comment by BenjaminRodenberg** [comment]
> Can we sync this with the latest develop to ensure that when we plan to merge this we would not have excessive conflicts?

Done.

File: cyprecice/cyprecice.pyx [hunk 1]
```diff
@@ -1090,10 +984,13 @@ cdef class Interface:
         size = _vertex_ids.size
         dimensions = self.get_dimensions()
         cdef np.ndarray[np.double_t, ndim=1] _values = np.empty(size * dimensions, dtype=np.double)
-        self.thisptr.readBlockVectorData (data_id, size, <const int*>_vertex_ids.data, <double*>_values.data)
+        if relative_read_time == None:
```
Comment (IshaanDesai) [inline cyprecice/cyprecice.pyx:987]:
```suggestion
        if not relative_read_time:
```

File: cyprecice/cyprecice.pyx [hunk 2]
```diff
@@ -1134,10 +1033,14 @@ cdef class Interface:
         """
         dimensions = self.get_dimensions()
         cdef np.ndarray[double, ndim=1] _value = np.empty(dimensions, dtype=np.double)
-        self.thisptr.readVectorData (data_id, vertex_id, <double*>_value.data)
+        if relative_read_time == None:
```
Comment (IshaanDesai) [inline cyprecice/cyprecice.pyx:1036]:
```suggestion
        if not relative_read_time:
```

File: cyprecice/cyprecice.pyx [hunk 3]
```diff
@@ -1176,10 +1081,14 @@ cdef class Interface:
         cdef np.ndarray[int, ndim=1] _vertex_ids = np.ascontiguousarray(vertex_ids, dtype=np.int32)
         size = _vertex_ids.size
         cdef np.ndarray[double, ndim=1] _values = np.empty(size, dtype=np.double)
-        self.thisptr.readBlockScalarData (data_id, size, <const int*>_vertex_ids.data, <double*>_values.data)
+        if relative_read_time == None:
```
Comment (IshaanDesai) [inline cyprecice/cyprecice.pyx:1084]:
```suggestion
        if not relative_read_time:
```

File: cyprecice/cyprecice.pyx [hunk 4]
```diff
@@ -1207,8 +1118,12 @@ cdef class Interface:
         >>> vertex_id = 5
         >>> value = interface.read_scalar_data(data_id, vertex_id)
         """
-        cdef double _value = 0
-        self.thisptr.readScalarData (data_id, vertex_id, _value)
+        cdef double _value
+        if relative_read_time == None:
```
Comment (IshaanDesai) [inline cyprecice/cyprecice.pyx:1122]:
```suggestion
        if not relative_read_time:
```

**Review by IshaanDesai — APPROVED** [review]
Looks good :+1: just a minor suggestion in the if statement to check if a relative time is provided or not.

**Comment by BenjaminRodenberg** [comment]
@IshaanDesai Thanks for the review and the suggestion! I would suggest to merge this PR as soon as https://github.com/precice/precice/pull/1366 is merged, because this PR is also a breaking one and relies on `develop-v3.0.0`. (just added the breaking tag)

**Pull request closed without merging**